core/aws-core/src/main/java/software/amazon/awssdk/awscore/eventstream/EventStreamInitialRequestInterceptor.java

```diff
@@ -0,0 +1,107 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.interceptor.Context.ModifyHttpRequest;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.core.internal.async.AsyncStreamPrepender;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.eventstream.HeaderValue;
+import software.amazon.eventstream.Message;
+
+/**
+ * An interceptor for event stream requests sent over RPC. This interceptor will prepend the initial request (i.e. the
+ * serialized request POJO) to the stream of events supplied by the caller.
+ */
+@SdkProtectedApi
+public class EventStreamInitialRequestInterceptor implements ExecutionInterceptor {
+    @Override
+    public SdkHttpRequest modifyHttpRequest(
+        ModifyHttpRequest context, ExecutionAttributes executionAttributes
+    ) {
+        if (!Boolean.TRUE.equals(executionAttributes.getAttribute(HAS_INITIAL_REQUEST_EVENT))) {
+            return context.httpRequest();
+        }
+
+        return context.httpRequest().toBuilder()
+            .removeHeader(CONTENT_TYPE)
+            .putHeader(CONTENT_TYPE, MIMETYPE_EVENT_STREAM)
+            .build();
+    }
+
+    @Override
+    public Optional<AsyncRequestBody> modifyAsyncHttpContent(
+        ModifyHttpRequest context, ExecutionAttributes executionAttributes
+    ) {
+        if (!Boolean.TRUE.equals(executionAttributes.getAttribute(HAS_INITIAL_REQUEST_EVENT))) {
+            return context.asyncRequestBody();
+        }
+
+        /*
+         * At this point in the request execution, the requestBody contains the serialized initial request,
+         * and the asyncRequestBody contains the event stream proper. We will prepend the former to the
+         * latter.
+         */
+        byte[] payload = getInitialRequestPayload(context);
+        String contentType = context.httpRequest().headers().get(CONTENT_TYPE).get(0);
+
+        Map<String, HeaderValue> initialRequestEventHeaders = new HashMap<>();
+        initialRequestEventHeaders.put(":message-type", HeaderValue.fromString("event"));
+        initialRequestEventHeaders.put(":event-type", HeaderValue.fromString("initial-request"));
+        initialRequestEventHeaders.put(":content-type", HeaderValue.fromString(contentType));
+
+        ByteBuffer initialRequest = new Message(initialRequestEventHeaders, payload).toByteBuffer();
+
+        Publisher<ByteBuffer> asyncRequestBody = context.asyncRequestBody()
+            .orElseThrow(() -> new IllegalStateException("This request should have an asyncRequestBody"));
```

Comment:
minor: can we indicate in the error message that it's because it's an event streaming request? Would help in debugging